userapi/src/main/java/fk/prof/userapi/api/cache/ZkLoadInfoStore.java

```diff
@@ -0,0 +1,312 @@
+import org.apache.zookeeper.data.Stat;
+
+import java.nio.charset.Charset;
+import java.time.LocalDateTime;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+/**
+ * Shared storage based on zookeeper to store the mapping profile -> ip:port and maintain resource utilization counters
+ * for each node in cluster.
+ * It is used by {@link ClusterAwareCache}.
+ *
+ * zookeeper node structure:
+ * /fkprof-userapi                                                                    (session namespace)
+ * /fkprof-userapi/nodesInfo/{ip:port} -> {@link NodeLoadInfo}                                (ephemeral)
+ * /fkprof-userapi/profilesLoadStatus/{profileName} -> {@link ProfileResidencyInfo}           (ephemeral)
+ *
+ * For resource utilization counter, currently only cached profile count is stored
+ *
+ * @see NodeLoadInfo
+ * @see ProfileResidencyInfo
+ *
+ * Created by gaurav.ashok on 01/08/17.
+ */
+class ZkLoadInfoStore {
+    private static final Logger logger = LoggerFactory.getLogger(ZkLoadInfoStore.class);
+    private static final String distributedLockPath = "/global_mutex";
+
+    private final String zkNodesInfoPath;
+    private final CuratorFramework zookeeper;
+    private final InterProcessSemaphoreMutex sharedMutex;
+
+    private final AtomicReference<ConnectionState> connectionState;
+    private final AtomicReference<LocalDateTime> lastZkLostTime;
+    private final AtomicBoolean recentlyZkConnectionLost;
+
+    private final byte[] myResidencyInfo;
+    private Supplier<List<AggregatedProfileNamingStrategy>> cachedProfiles;
+
+    private Counter lockAcquireTimeoutCounter = Util.counter("zk.lock.timeouts");
+    private Counter onReconnectFailures = Util.counter("zk.onreconnect.failures");
+
+    ZkLoadInfoStore(CuratorFramework zookeeper, String myIp, int port, Supplier<List<AggregatedProfileNamingStrategy>> cachedProfiles) {
+        this.zookeeper = zookeeper;
+
+        this.zkNodesInfoPath = "/nodesInfo/" + myIp + ":" + port;
+        this.sharedMutex = new InterProcessSemaphoreMutex(zookeeper, distributedLockPath);
+        this.lastZkLostTime = new AtomicReference<>(LocalDateTime.MIN);
+        this.recentlyZkConnectionLost = new AtomicBoolean(false);
+
+        this.myResidencyInfo = ProfileResidencyInfo.newBuilder().setIp(myIp).setPort(port).build().toByteArray();
+
+        this.zookeeper.getConnectionStateListenable().addListener(this::zkStateChangeListener,
+            Executors.newSingleThreadExecutor(new ThreadFactoryBuilder().setNameFormat("curator-state-listener").build()));
+
+        this.cachedProfiles = cachedProfiles;
+
+        this.connectionState = new AtomicReference<>(
+            zookeeper.getZookeeperClient().isConnected() ? ConnectionState.Connected : ConnectionState.Disconnected);
+    }
+
+    boolean ensureBasePathExists() throws Exception {
```

Comment:
return void
a better name would be initialized or init

Reply:
Made the return value void